### PR TITLE
[Prototype] Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time

### PR DESCRIPTION
**Pull request by voznesenskym**
cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -803,10 +782,34 @@ def flat_fn(*flat_args):
...
+            # aot_function is the "public" entrypoint, so we need to process here
+            # For internal entrypoint with already populated fake tensors, see aot_function_simplified
+
+            def process_inputs(flat_args):
+                if config.use_fake_tensor:
+                    def convert(idx, x):
+                        if not isinstance(x, torch.Tensor):
+                            return x
+                        if isinstance(x, torch._subclasses.fake_tensor.FakeTensor):
+                            return x
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:798]: This looks questionable. If an argument is already a fake tensor, it's unlikely to be consistent with the freshly allocated fake mode. Which means you'd probably get an error if you tried to actually use it. Better to not support this case.

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -865,7 +868,18 @@ def forward(self, *args, **kwargs):
     return AOTModule()
 
 
-def aot_module_simplified(mod: nn.Module, *top_args, **top_kwargs) -> nn.Module:
+def aot_module_simplified(
+        mod: nn.Module,
+        inputs,
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:873]: A comment saying what the acceptable `inputs` here would be good.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -865,7 +868,18 @@ def forward(self, *args, **kwargs):
...
+        fw_compiler: Callable,
+        bw_compiler: Optional[Callable] = None,
+        partition_fn: Callable = default_partition,
+        decompositions: Optional[Dict] = None,
+        hasher_type=None,
+        static_argnums=None,
+        fake_mode=None,
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:880]: What the heck is this kwarg, you ignore it immediately after?

File: functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -886,6 +899,37 @@ def aot_module_simplified(mod: nn.Module, *top_args, **top_kwargs) -> nn.Module:
     params_flat = tuple(params_flat)
     params_len = len(params_flat)
 
+    # TODO(voz): Pull up to dynamo
+    # See [Real vs Fake Parms] below
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:903]: Params

File: functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -886,6 +899,37 @@ def aot_module_simplified(mod: nn.Module, *top_args, **top_kwargs) -> nn.Module:
...
+    fake_mode = None
+
+    def fakify_params_and_buffers(flat_args):
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:906]: You sort of cover this in the Note below, but (1) honestly the note should get pulled up earlier and (2) a direct discussion about why inputs is always fake, but not always params/buffers would be helpful for readers

File: functorch/_src/aot_autograd.py [hunk 6]
```diff
@@ -908,56 +952,51 @@ def functional_call(*args, **kwargs):
...
+    #
+    # So, we have a few ways of handling it
+    # 1) Dynamo passes only real params in, we fakify at the aot level (this current state)
+    # 2) Dynamo passes only fake params in, but we rewrite the def forward(*args): to def forward(*parms, *args):
+    #    (Not bad, but very annoying w/r/t changing dynamo's calling convention and contracts. Also, not having to close
+    #    over params is nice, and better w/r/t lifecycles and mutations)
+    # 3) Dynamo passes both real and fake tensor in (Not a bad future, but really a stopgap to 2)
+    #
+    # (2) is the correct long term direction, and we are at (1) for now, (3) can be done to make things a little more streamlined
+    # w/r/t where fake tensor conversion happens.
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:987]: This note doesn't really explain why you would prefer fake or real in the first place. Clearly runtime needs real. But compile time sometimes needs real too, and it's worth spelling that out explicitly here. From the description here it's not clear what `_create_aot_dispatcher_function` means

File: test/dynamo/test_minifier.py [hunk 7]
```diff
@@ -96,7 +96,7 @@ def inner(x):
         )
 
         (test_proc, _, repro_proc), _ = self._run_full_test(
-            run_code, "dynamo", repro_level, backend_code
+            run_code, "dynamo", repro_level, backend_code, False
```
Comment (ezyang) [inline test/dynamo/test_minifier.py:99]: this is some pretty bad boolean blindness. Kwargify it please.

File: test/inductor/test_torchinductor.py [hunk 8]
```diff
@@ -238,6 +238,7 @@ def check_model(
     assert_equal=True,
     check_gradient=False,
 ):
+    # breakpoint()
```
Comment (ezyang) [inline test/inductor/test_torchinductor.py:241]: delete

File: torch/_inductor/config.py [hunk 9]
```diff
@@ -93,13 +93,13 @@ class cpp:
...
-        "g++-10",
-        "clang++",
+        # None,  # download gcc12 from conda-forge if conda is installed
+        # "g++-12",
+        # "g++-11",
+        # "g++-10",
+        # "clang++",
         "g++",
-        "g++.par",
+        # "g++.par",
```
Comment (ezyang) [inline torch/_inductor/config.py:102]: ??

File: torch/_dynamo/eval_frame.py [hunk 10]
```diff
@@ -326,15 +326,26 @@ def __init__(self, backend=None):
...
+    def __call__(self, gm: torch.fx.GraphModule, fake_inputs, real_inputs):
 
         self.restore = checkpoint_params(gm)
-        self.original_example_inputs = clone_inputs(example_inputs)
+        self.original_example_inputs_real = clone_inputs(real_inputs)
+        self.original_example_inputs_fake = clone_inputs(fake_inputs)
         self.gm = gm
         copy_gm = copy.deepcopy(self.gm)
-        self.candidate = self.backend(copy_gm, self.original_example_inputs)
+        if config.fake_tensor_propagation:
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:340]: For later: let's please get rid of this feature flag lol

File: torch/_dynamo/eval_frame.py [hunk 11]
```diff
@@ -326,15 +326,26 @@ def __init__(self, backend=None):
...
+            # This is a *huge hack* to get around the fact that
+            # AOTAutograd mutation analysis only works with taking real inputs atm
+            # it has its own weird and brittle fakification story that is solved on
+            # the symbolic shapes branch.
+            # this wil be removed ASAP.
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:345]: Is this comment still up to date?

File: torch/_dynamo/optimizations/backends.py [hunk 12]
```diff
@@ -55,7 +55,7 @@ def inner(model, example_inputs=None, **kwargs):
             raise
         except Exception:
             log.exception(f"{fn.__name__} error")
-            return None
+            raise
```
Comment (ezyang) [inline torch/_dynamo/optimizations/backends.py:58]: Do you want to permanently land this?

File: torch/_dynamo/output_graph.py [hunk 13]
```diff
@@ -458,20 +458,48 @@ def call_user_compiler(self, gm):
                 else ""
             )
             _step_logger()(logging.INFO, f"calling compiler function {name}")
-            compiled_fn = self.compiler_fn(gm, self.example_inputs())
+            from inspect import signature
+
+            try: 
+                print("COMPILER FN", self.compiler_fn)
+                print("FAKE?", config.fake_tensor_propagation)
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:465]: remove

File: torch/_dynamo/test_minifier_common.py [hunk 14]
```diff
@@ -66,6 +66,7 @@ def _run_test_code(self, code):
     def _inject_code(self, patch_code, filename):
         patch_code = f"""\
 {patch_code}
+torch._dynamo.config.fake_tensor_propagation = False
```
Comment (ezyang) [inline torch/_dynamo/test_minifier_common.py:69]: This seems morally wrong though; you want dynamo to be running as similar to the original configuration as possible, so that you are likely to get the bug. If the bug being minified is related to incorrect striding, turning off fake tensor propagation may cause the bug to go away.

File: torch/_dynamo/test_minifier_common.py [hunk 15]
```diff
@@ -106,14 +107,17 @@ def _run_repro(self, patch_code, repro_dir):
...
         return f"""\
 import torch
 import torch._dynamo
 {patch_code}
 torch._dynamo.config.repro_after = "{repro_after}"
 torch._dynamo.config.repro_level = {repro_level}
 torch._dynamo.config.debug_dir_root = "{self.DEBUG_DIR}"
+# Currently, when called with fake tensors, test_relu_accuracy_error raises a DataDependentOutputException
+# instead of the desired BadAccuracyDetected
+torch._dynamo.config.fake_tensor_propagation = {fake_tensor}
```
Comment (ezyang) [inline torch/_dynamo/test_minifier_common.py:120]: looks questionable...

File: torch/_dynamo/utils.py [hunk 16]
```diff
@@ -397,6 +397,9 @@ def clone_tensor(x):
 
 def clone_input(x):
     """copy while preserving strides"""
+    if isinstance(x, torch._subclasses.FakeTensor):
+        # this func fails on fake tensors in __torch_dispatch__
+        return x
```
Comment (ezyang) [inline torch/_dynamo/utils.py:402]: nuh uh, this is wrong. For fake tensor it may be acceptable to just .detach(), as a data copy is not necessary; it depends on whether or not you want the storage to be distinct or not. But better to not overload `clone_input` to work on both fake and real, IMO

File: torch/_inductor/overrides.py [hunk 17]
```diff
@@ -449,7 +449,11 @@ def fuse_fx(gm: torch.fx.GraphModule, example_inputs):
     gm = fuse_conv_bn(gm)
     # For binary fusion, we need to check inputs info to make sure
     # the binary inputs have same tensor info(device, dtype, and layout).
-    ShapeProp(gm).propagate(*example_inputs)
+    
+    # Doesn't work with fake tensors, do not land like this
+    # ShapeProp(gm).propagate(*example_inputs)
+    # Doesn't work with fake tensors, do not land like this
```
Comment (ezyang) [inline torch/_inductor/overrides.py:455]: TODO

File: torch/_inductor/compile_fx.py [hunk 18]
```diff
@@ -355,6 +357,23 @@ def compile_fx(
...
+                fake_mode = input.fake_mode
+            else:
+                assert fake_mode == input.fake_mode
+
+    # if fake_mode:
+        # NOTE: This *will* create guards - we are missng
+        # the shape env guard supression logic here.
+        # We need to add it once its pushed up.
+        # Don't land this without that logic. 
+        # model_ = deepcopy_to_fake_tensor(model_, fake_mode)
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:376]: TODO I guess this part is WIP

File: torch/_dynamo/optimizations/analysis.py [hunk 19]
```diff
@@ -138,11 +138,26 @@ def _wrap_to_fake_tensor(t, *, f_mode):
...
-            shape_env=ShapeEnv() if config.dynamic_shapes else None
-        )
+        fake_mode = None
+        flat_inputs, _ = tree_flatten(example_inputs)
+        for input in flat_inputs:
+            if isinstance(input, torch._subclasses.FakeTensor):
+                if fake_mode is None:
+                    fake_mode = input.fake_mode
+                else:
+                    assert fake_mode == input.fake_mode
```
Comment (ezyang) [inline torch/_dynamo/optimizations/analysis.py:148]: You reimplemented this in a bunch of places, time to factor it into a function

File: torch/_dynamo/variables/builder.py [hunk 20]
```diff
@@ -120,6 +122,14 @@ def load(self, tx):
     def get_examples(self):
         return [self.example]
 
+    def get_fake_examples(self):
+        assert config.fake_tensor_propagation
+        if self.fake_tensor is not None:
+            assert isinstance(
+                self.fake_tensor, torch._subclasses.fake_tensor.FakeTensor
+            )
+        return [self.fake_tensor]
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:131]: so... what if it's None? Is it even meaningful to return [None] here?

File: torch/_dynamo/variables/builder.py [hunk 21]
```diff
@@ -564,6 +574,12 @@ def wrap_tensor(self, value: torch.Tensor):
                     guards=self.make_guards(GuardBuilder.TENSOR_MATCH),
                     should_specialize=self.tensor_should_specialize(),
                 )
+            if graph_arg and config.fake_tensor_propagation:
+                fake_tensor_value = tensor_variable.proxy.node.meta["example_value"]
+                if isinstance(
+                    fake_tensor_value, torch._subclasses.fake_tensor.FakeTensor
+                ):
+                    graph_arg.fake_tensor = fake_tensor_value
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:582]: @jansel lmk if you think this is the best way to get the fake tensor onto Grapharg

File: functorch/_src/aot_autograd.py (hunk 3 above)
Comment (voznesenskym) [reply functorch/_src/aot_autograd.py:880]: Leftover

File: functorch/_src/aot_autograd.py (hunk 4 above)
Comment (voznesenskym) [reply functorch/_src/aot_autograd.py:903]: Parmesan?

File: torch/_dynamo/output_graph.py (hunk 13 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:465]: Not yet :)

File: torch/_dynamo/utils.py (hunk 16 above)
Comment (voznesenskym) [reply torch/_dynamo/utils.py:402]: Yeah this is a placeholder while I get the rest of the stuff in shape. Will ping you about it in chat.

File: torch/_dynamo/variables/builder.py (hunk 20 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:131]: It's mirroring the behavior of the func above it, get_examples

File: torch/_inductor/compile_fx.py (hunk 18 above)
Comment (voznesenskym) [reply torch/_inductor/compile_fx.py:376]: Whole PR is WIP :) but yes this is for sure

File: torch/_inductor/config.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_inductor/config.py:102]: Was just testing stuff, won't land

File: torch/_inductor/overrides.py (hunk 17 above)
Comment (voznesenskym) [reply torch/_inductor/overrides.py:455]: This is more than a todo, this is a DONOTLAND

File: torch/_dynamo/eval_frame.py [hunk 22]
```diff
@@ -326,15 +326,26 @@ def __init__(self, backend=None):
...
+        if config.fake_tensor_propagation:
+            return clone_inputs(self.original_example_inputs_fake)
+        return clone_inputs(self.original_example_inputs_real)
 
-    def __call__(self, gm: torch.fx.GraphModule, example_inputs):
+    def __call__(self, gm: torch.fx.GraphModule, fake_inputs, real_inputs):
 
         self.restore = checkpoint_params(gm)
-        self.original_example_inputs = clone_inputs(example_inputs)
+        self.original_example_inputs_real = clone_inputs(real_inputs)
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:336]:
```suggestion
        assert config.verify_correctness
        self.original_example_inputs_real = clone_inputs(real_inputs)
```
This clone will be expensive, so want to make sure we only hit it with this config flag set.

File: torch/_dynamo/variables/builder.py (hunk 21 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py:582]: Seems fine, might be cleaner to do it closer to when it is constructed.

File: torch/_dynamo/debug_utils.py [hunk 23]
```diff
@@ -862,7 +868,12 @@ def debug_wrapper(gm, example_inputs, **kwargs):
                     )
                     raise
         else:
-            compiled_gm = compiler_fn(gm, example_inputs, **kwargs)
+            if "real_inputs" in signature(compiler_fn).parameters:
```
Comment (jansel) [inline torch/_dynamo/debug_utils.py:871]: As far as I can tell the only place where `real_inputs` is used is to support `config.verify_correctness` mode (aka WrapperBackend), which just adds a correctness check after each `compile_fn` call.  This mode is largely unused.

This seems like a lot of complexity for something rather simple.  I'd suggest:

1) make `config.verify_correctness` require `not config.fake_tensor_propagation` (so we don't need `real_inputs=` anymore)

2) or, delete `WrapperBackend` and reimplement `config.verify_correctness` directly at the call site to `compiler_fn`.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/debug_utils.py (hunk 23 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:871]: Sgtm!

File: torch/_dynamo/eval_frame.py (hunk 22 above)
Comment (voznesenskym) [reply torch/_dynamo/eval_frame.py:336]: Good call!

**Comment by voznesenskym** [comment]
@ezyang @jansel thanks for taking a look. This is just a prototype of what we were discussing, so it's very subject to change, lots of tests still don't pass, I need to get ShapeProp working, etc.

File: torch/_dynamo/variables/builder.py (hunk 20 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:131]: But in the code above, it seems to be that self.example is guarantee to not be None?

File: torch/_dynamo/optimizations/distributed.py [hunk 24]
```diff
@@ -236,42 +247,69 @@ def forward(self, *args):
...
                     # modify the currently running FX graph
                     # maybe this isn't sound in general, but only changing the target of a node might be ok?
                     if n.op == "call_module":
-                        submod = self.fetch_attr(n.target)
-                        log.debug(f"\n---{n.target} graph---\n" + str(submod.graph))
-                        compiled_submod = self.compile_submod(submod, args, kwargs)
+                        real_mod = self.fetch_attr(n.target)
+                        if fake_mode:
+                            from ..utils import deepcopy_to_fake_tensor
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
```
Comment (wconstab) [inline torch/_dynamo/optimizations/distributed.py:297]: sigh.  not a fan of all these deepcopy to fake tensor things.  but i don't think i had any issues with them for DDP, only FSDP.

File: torch/_dynamo/optimizations/distributed.py [hunk 25]
```diff
@@ -236,42 +247,69 @@ def forward(self, *args):
                         return x
 
                 unwrap_singleton_tuple = False
-                for sn in submod.graph.nodes:
+                for sn in real_mod.graph.nodes:
```
Comment (wconstab) [inline torch/_dynamo/optimizations/distributed.py:250]: would it be better to zip the 2 nodes lists and assert one of them matches the other instead of duplicating the traversal separately?

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (voznesenskym) [reply functorch/_src/aot_autograd.py:798]: Fwiw this is copied fro the old functionality, but let's see if we can make things better than we found them.

File: torch/_dynamo/optimizations/distributed.py [hunk 26]
```diff
@@ -236,42 +247,69 @@ def forward(self, *args):
...
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
+                    fake_mode = None
+                    for tensor in args:
+                        if isinstance(tensor, torch._subclasses.FakeTensor):
+                            fake_mode = tensor.fake_mode
+                            break
+                    new_args = []
+                    if fake_mode:
+                        for arg in args:
+                            if isinstance(arg, torch.Tensor) and not isinstance(arg, torch._subclasses.FakeTensor):
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:280]: Sure, this is prototype code, we can do this way cleaner.

File: torch/_dynamo/optimizations/distributed.py (hunk 24 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:297]: Hmm, what issues?

File: torch/_dynamo/optimizations/distributed.py (hunk 25 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/distributed.py:250]: Eh, I think this is all just duplicated cause I havent pulled it into a function yet

File: torch/_dynamo/optimizations/distributed.py [hunk 27]
```diff
@@ -236,42 +247,69 @@ def forward(self, *args):
...
+                            from ..utils import deepcopy_to_fake_tensor
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
+                        else:
+                            fake_submod = real_mod
+                            pass
+                        log.debug(f"\n---{n.target} graph---\n" + str(fake_submod.graph))
+                        compiled_submod = self.compile_submod(fake_submod, real_mod, new_args, kwargs)
                         self.module.delete_submodule(n.target)
                         n.target = "compiled_" + n.target
                         self.module.add_submodule(n.target, compiled_submod)
```
Comment (ezyang) [inline torch/_dynamo/optimizations/distributed.py:305]: manually call (and return) `fake_submod(*args, **kwargs)` (to continue propagating fake tensors). DO NOT modify the input FX graph, this is no longer necessary

File: torch/_dynamo/optimizations/distributed.py (hunk 24 above)
Comment (wconstab) [reply torch/_dynamo/optimizations/distributed.py:280]: well, FSDP modules were not deep-copyable but it turned out to not be an issue since FSDP code itself is probably not dynamo friendly in other ways and we just skip it all now.

i don't think this will cause issues.

**Comment by wconstab** [comment]
we may want to run a sweep over DDP benchmark models to verify no new errors/exceptions due to the changes

past sweep results: https://docs.google.com/spreadsheets/d/1XBvFnKH8H3mWDKrcWxubjxEzBCn6HMMtnQzfSST4u6s/edit#gid=924630697

command to run the sweep  (probably just run the inductor one and compare to last time above)

ddp_sweep.sh:
```
for SUITE in "torchbench" "huggingface" "timm_models"
do
    SCRIPT="benchmarks/dynamo/${SUITE}.py"
    python $SCRIPT --training --accuracy --backend eager --ddp --no-optimize-ddp --output ${SUITE}_eager_ddp_no_optimize.csv
    python $SCRIPT --training --accuracy --backend aot_eager --ddp --no-optimize-ddp --output ${SUITE}_aot_eager_ddp_no_optimize.csv
    python $SCRIPT --training --accuracy --backend aot_eager --ddp --output ${SUITE}_aot_eager_ddp.csv
    python $SCRIPT --training --accuracy --inductor --ddp --no-optimize-ddp --output ${SUITE}_inductor_ddp_no_optimize.csv
    python $SCRIPT --training --accuracy --inductor --ddp --output ${SUITE}_inductor_ddp.csv
done
``` 

We may also want to kick off a perf benchmark run @davidberard98 - is the command handy?  probably just one model is fine as a smoke test but idk.

**Comment by davidberard98** [comment]
> We may also want to kick off a perf benchmark run @davidberard98 - is the command handy? probably just one model is fine as a smoke test but idk.

From torchbench directory (https://github.com/pytorch/benchmark) - need to have `torchvision, transformers, timm` installed:

```
$ python userbenchmark/ddp_experiments/__init__.py --job_dir /fsx/path/to/log/location --partition train
```

^ record the job_id reported to stdout

then to gather results:
```
$ python userbenchmark/ddp_experiments/parse_ddp.py --results_dir /fsx/path/to/log/location --job_id [JOB_DIR]
```

**Comment by ezyang** [comment]
I'm not rereviewing this until asked. Unsubbing for now

File: functorch/_src/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:799]: This isn't copied. The old code doesn't explicitly test for FakeTensor.

**Comment by ezyang** [comment]
if you avoid force pushing that will make it easier for me to collaborate on https://github.com/pytorch/pytorch/pull/89636 thanks!

File: torch/_dynamo/output_graph.py [hunk 28]
```diff
@@ -458,20 +461,59 @@ def call_user_compiler(self, gm):
                 else ""
             )
             _step_logger()(logging.INFO, f"calling compiler function {name}")
-            compiled_fn = self.compiler_fn(gm, self.example_inputs())
+
+            backend_correct = True
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:465]: I'm going to factor this (moving wrapper compiler inline) out into a standalone PR

Comment (ezyang) [reply torch/_dynamo/output_graph.py:465]: I also notice that you did some unrelated refactoring while you're in the code; I'm going to slavishly follow the original implementation, and you should feel free to reapply your refactor if you still want it.

Comment (ezyang) [reply torch/_dynamo/output_graph.py:465]: https://github.com/pytorch/pytorch/pull/89662

It doesn't pull the trigger on the API change but should make it simple to do so when we actually need it.

File: functorch/_src/aot_autograd.py [hunk 29]
```diff
@@ -908,56 +960,36 @@ def functional_call(*args, **kwargs):
...
+        fw_compiler=fw_compiler,
+        bw_compiler=bw_compiler,
+        partition_fn=partition_fn,
+        decompositions=decompositions,
+        num_params_buffers=params_len,
+    )
 
-        return new_func
+    fake_flat_tensor_params.extend(inputs)
+    aot_dispatcher_function = _create_aot_dispatcher_function(functional_call, fake_flat_tensor_params, aot_config, fake_mode)
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:976]: This is a good refactor, I extracted it out into a standalone commit for ease of review: https://github.com/pytorch/pytorch/pull/89664

File: functorch/_src/aot_autograd.py [hunk 30]
```diff
@@ -865,7 +869,16 @@ def forward(self, *args, **kwargs):
...
-def aot_module_simplified(mod: nn.Module, *top_args, **top_kwargs) -> nn.Module:
+def aot_module_simplified(
+        mod: nn.Module,
+        inputs: List[torch.Tensor],
+        fw_compiler: Callable,
+        bw_compiler: Optional[Callable] = None,
+        partition_fn: Callable = default_partition,
+        decompositions: Optional[Dict] = None,
+        hasher_type=None,
+        static_argnums=None,
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:880]: This is also a good refactor, also factored out for ease of review: https://github.com/pytorch/pytorch/pull/89666

File: functorch/_src/aot_autograd.py [hunk 31]
```diff
@@ -865,7 +869,16 @@ def forward(self, *args, **kwargs):
     return AOTModule()
 
 
-def aot_module_simplified(mod: nn.Module, *top_args, **top_kwargs) -> nn.Module:
+def aot_module_simplified(
+        mod: nn.Module,
+        inputs: List[torch.Tensor],
```
Comment (ezyang) [inline functorch/_src/aot_autograd.py:874]: The reason you added the inputs is so that you can extract out a fake tensor mode from them. But then... why not just pass in a fake tensor mode then lol.

File: torch/_dynamo/optimizations/backends.py [hunk 32]
```diff
@@ -561,7 +560,7 @@ def _wrapped_bw_compiler(*args, **kwargs):
 
     from .. import disable
 
-    return aot_module_simplified(subgraph.model, **kwargs)
+    return aot_module_simplified(subgraph.model, subgraph.example_inputs, **kwargs)
```
Comment (ezyang) [inline torch/_dynamo/optimizations/backends.py:563]: I think this is the only call site that actually matters
